### PR TITLE
Loki: Introduce `$__auto` range variable for metric queries

### LINKED ISSUE #71743: Loki: Create [auto] range to sync step/timerange span with range in queries
Discovered during working on https://github.com/grafana/grafana/issues/58337

**Range [auto] that would use step interval for range queries and full range interval for instant queries:**
In Loki, you need to set “step” usually at 2 places - in  [range] and in step. However, most of the time, you want [range] to be:
- Same as step for range queries
- Same as time range for instant queries

After discussion with the Loki team, we agreed that we would like to introduce [auto] range, which would default to values mentioned above. 

Moreover, in the future, Loki plans to support metric queries without range, or with [auto] range.

Here is the design doc: https://docs.google.com/document/d/1giA6oxb_gyAJYvI2b0qL1XACgVrsS1HsiN0ZzCDze7o/edit#heading=h.41dvitd003pb

### PR DESCRIPTION
**Pull request by ivanahuckova**
**What is this feature?**

We are introducing `$__auto` variable for Loki. This is similar for what we have done in Prometheus with `$__rate_interval`. 

**Why do we need this feature?**

In Loki, you need to set “step” usually at 2 places - in  [range] and in step. However, most of the time, you want [range] to be:
- Same as step for range queries
- Same as time range for instant queries

After discussion with the Loki team, we agreed that we would like to introduce $__auto range, which would default to values mentioned above. $__auto variable naming is consistent with Grafana variable syntax and it would be clear that is is variable. 

**Who is this feature for?**
For Loki users

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/71743

**How to test this:**
1. Initialise Loki with `make devenv sources=loki`
2. Open Explore and run queries (both types - Instant and Range ) that include `[$__auto]` e.g. `rate({job="grafana"}[$__auto)` and ensure that queries are successfully executed.
3. Check in Code editor that $__auto is correctly validated - no red squiggly line and that it is offered as autocomplete suggestion
4. Check in Query builder that when you select `Rate` operation, it will give you $__auto as default value. Here also hover over "Explain this Operation" to see if $__auto has more detailed explanation. 
5. Check that in Kick start your query metric queries use $__auto and not $__interval
6. Check in label browser that when you click "Show logs rate", it uses $__auto

https://github.com/grafana/grafana/assets/30407135/ddc34b26-bb11-4dc3-baf8-1d2523d23816

**Special notes for your reviewer:**

We have decided to remove `$__interval` and `$__range` variables from the suggestions because `$__auto` replaces both and we don't want to confuse users with multiple options.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: pkg/tsdb/loki/parse_query.go [hunk 1]
```diff
@@ -42,6 +46,15 @@ func interpolateVariables(expr string, interval time.Duration, timeRange time.Du
 	expr = strings.ReplaceAll(expr, varRangeMs, rangeMsText)
 	expr = strings.ReplaceAll(expr, varRangeS, rangeSText)
 	expr = strings.ReplaceAll(expr, varRange, rangeSText+"s")
+	if queryType == dataquery.LokiQueryTypeInstant {
+	// For instant we want to replace with $__range
```
Comment (ivanahuckova) [inline pkg/tsdb/loki/parse_query.go:50]:
```suggestion
```

File: docs/sources/datasources/loki/template-variables/index.md [hunk 2]
```diff
@@ -47,11 +47,11 @@ You can use this variable type to specify any number of key/value filters, and G
 
 For more information, refer to [Add ad hoc filters][add-template-variables-add-ad-hoc-filters].
 
-## Use interval and range variables
+## Use $\_\_auto variable for Loki metric queries
 
-You can use some global built-in variables in query variables: `$__interval`, `$__interval_ms`, `$__range`, `$__range_s`, and `$__range_ms`.
+Consider using the `$__auto` variable in your Loki metric queries, which will be automatically substituted with the [step value](https://grafana.com/docs/grafana/next/datasources/loki/query-editor/#options) for range queries, and with the selected time range's value (computed from the starting and ending times) for instant queries.
```
Comment (lwandz13) [inline docs/sources/datasources/loki/template-variables/index.md:52]:
```suggestion
Consider using the `$__auto` variable in your Loki metric queries, which will automatically be substituted with the [step value](https://grafana.com/docs/grafana/next/datasources/loki/query-editor/#options) for range queries, and with the selected time range's value (computed from the starting and ending times) for instant queries.
```

**Review by lwandz13 — APPROVED** [review]
one little suggestion - looks good!

File: public/app/plugins/datasource/loki/querybuilder/operationUtils.ts [hunk 3]
```diff
@@ -53,8 +53,8 @@ export function createRangeOperation(name: string, isRangeOperationWithGrouping?
     explainHandler: (op, def) => {
       let opDocs = FUNCTIONS.find((x) => x.insertText === op.id)?.documentation ?? '';
 
-      if (op.params[0] === '$__interval') {
-        return `${opDocs} \`$__interval\` is a variable that will be replaced with the [calculated interval](https://grafana.com/docs/grafana/latest/dashboards/variables/add-template-variables/#__interval) based on the time range and width of the graph. In Dashboards, you can affect the interval variable using **Max data points** and **Min interval**. You can find these options under **Query options** right of the data source select dropdown.`;
+      if (op.params[0] === '$__auto') {
+        return `${opDocs} \`$__auto\` is a variable that will be replaced with the [value of step](https://grafana.com/docs/grafana/next/datasources/loki/query-editor/#options) for range queries and with the value of selected time range (calculated to - from) for instant queries.`;
```
Comment (matyax) [inline public/app/plugins/datasource/loki/querybuilder/operationUtils.ts:57]:
```suggestion
        return `${opDocs} \`$__auto\` is a variable that will be replaced with the [value of step](https://grafana.com/docs/grafana/next/datasources/loki/query-editor/#options) for range queries and with the value of the selected time range (calculated to - from) for instant queries.`;
```

**Review by matyax — APPROVED** [review]
Looks great! 👏

**Review by gabor — APPROVED** [review]
very nice 👍

**Pull request merged**